File: py/webnotes/widgets/form/assign_to.py
# Copyright (c) 2012 Web Notes Technologies Pvt Ltd (http://erpnext.com)
# 
# MIT License (MIT)
# 
# Permission is hereby granted, free of charge, to any person obtaining a 
# copy of this software and associated documentation files (the "Software"), 
# to deal in the Software without restriction, including without limitation 
# the rights to use, copy, modify, merge, publish, distribute, sublicense, 
# and/or sell copies of the Software, and to permit persons to whom the 
# Software is furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in 
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
# INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
# PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
# CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
# OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
# 

from __future__ import unicode_literals
"""assign/unassign to ToDo"""

import webnotes

@webnotes.whitelist()
def get(args=None):
	"""get assigned to"""
	if not args:
		args = webnotes.form_dict
	return webnotes.conn.sql("""select owner from `tabToDo`
		where reference_type=%(doctype)s and reference_name=%(name)s
		order by modified desc limit 5""", args, as_dict=1)
		
@webnotes.whitelist()
def add(args=None):
	"""add in someone's to do list"""
	if not args:
		args = webnotes.form_dict
	
	if webnotes.conn.sql("""select owner from `tabToDo`
		where reference_type=%(doctype)s and reference_name=%(name)s
		and owner=%(assign_to)s""", args):
		webnotes.msgprint("Already in todo")
		return
	else:
		from webnotes.model.doc import Document
		from webnotes.utils import nowdate
		
		d = Document("ToDo")
		d.owner = args['assign_to']
		d.reference_type = args['doctype']
		d.reference_name = args['name']
		d.description = args['description']
		d.priority = args.get('priority', 'Medium')
		d.date = args.get('date', nowdate())
		d.assigned_by = args.get('assigned_by', webnotes.user.name)
		d.save(1)

	# notify
	notify_assignment(d.assigned_by, d.owner, d.reference_type, d.reference_name, action='ASSIGN', notify=args.get('notify'))
		
	# update feeed
	try:
		import home
		from webnotes.utils import get_fullname
		home.make_feed('Assignment', d.reference_type, d.reference_name, webnotes.session['user'],
			'[%s] Assigned to %s' % (d.priority, get_fullname(d.owner)), '#C78F58')
	except ImportError, e:
		pass
	
	
	return get(args)

@webnotes.whitelist()
def remove(args=None):
	"""remove from todo"""
	if not args:
		args = webnotes.form_dict
	
	res = webnotes.conn.sql("""\
		select assigned_by, owner, reference_type, reference_name from `tabToDo`
		where reference_type=%(doctype)s and reference_name=%(name)s
		and owner=%(assign_to)s""", args)

	webnotes.conn.sql("""delete from `tabToDo`
		where reference_type=%(doctype)s and reference_name=%(name)s
		and owner=%(assign_to)s""", args)

	if res and res[0]: notify_assignment(res[0][0], res[0][1], res[0][2], res[0][3])

	return get(args)


def notify_assignment(assigned_by, owner, doc_type, doc_name, action='CLOSE', notify=0):
	"""
		Notify assignee that there is a change in assignment
	"""
	if not (assigned_by and owner and doc_type and doc_name): return

	from webnotes.boot import get_fullnames
	user_info = get_fullnames()

	# Search for email address in description -- i.e. assignee
	assignment = """<a href="#!Form/%s/%s">%s: %s</a>""" % (doc_type, doc_name,
			doc_type, doc_name)
	if action=='CLOSE':
		if owner == webnotes.session.get('user'):
			arg = {
				'contact': assigned_by,
				'txt': "The task %s, that you assigned to %s, has been \
					closed." % (assignment,
						user_info.get(owner, {}).get('fullname'))
			}
		else:
			arg = {
				'contact': assigned_by,
				'txt': "The task %s, that you assigned to %s, \
					has been closed	by %s." % (assignment,
					user_info.get(owner, {}).get('fullname'),
					user_info.get(webnotes.session.get('user'),
						{}).get('fullname'))
			}
	else:
		arg = {
			'contact': owner,
			'txt': "A new task, %s, has been assigned to you by %s." \
				% (assignment,
				user_info.get(webnotes.session.get('user'), {}).get('fullname')),
			'notify': notify
		}
	from utilities.page.messages import messages
	import json
	messages.post(json.dumps(arg))

File: public/js/legacy/wn/widgets/form/assign_to.js
// Copyright (c) 2012 Web Notes Technologies Pvt Ltd (http://erpnext.com)
// 
// MIT License (MIT)
// 
// Permission is hereby granted, free of charge, to any person obtaining a 
// copy of this software and associated documentation files (the "Software"), 
// to deal in the Software without restriction, including without limitation 
// the rights to use, copy, modify, merge, publish, distribute, sublicense, 
// and/or sell copies of the Software, and to permit persons to whom the 
// Software is furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in 
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
// OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
// 

// assign to is lined to todo
// refresh - load todos
// create - new todo
// delete to do
wn.widgets.form.sidebar.AssignTo = Class.extend({
	init: function(parent, sidebar, doctype, docname) {
		var me = this;
		this.doctype = doctype;
		this.name = docname;
		this.wrapper = $a(parent, 'div', 'sidebar-comment-wrapper');
		this.body = $a(this.wrapper, 'div');
		this.add_btn = $btn($a(this.wrapper, 'div', 'sidebar-comment-message'), 'Assign', 
			function() {
				me.add();
			})
		
		this.refresh();
	},
	refresh: function() {
		var me = this;
		$c('webnotes.widgets.form.assign_to.get', {
			doctype: me.doctype,
			name: me.name
		}, function(r,rt) {
			me.render(r.message)
		})
	},
	render: function(d) {
		var me = this;
		$(this.body).empty();
		if(this.dialog) {
			this.dialog.hide();			
		}
		
		for(var i=0; i<d.length; i++) {
			$(this.body).append(repl('<div>%(owner)s \
				<a class="close" href="#" data-owner="%(owner)s">&#215</a></div>', d[i]))
		}

		// set remove
		$(this.body).find('a.close').click(function() {
			$c('webnotes.widgets.form.assign_to.remove', {
				doctype: me.doctype,
				name: me.name,
				assign_to: $(this).attr('data-owner')		
			}, function(r,rt) {me.render(r.message);});
			return false;
		});
	},
	add: function() {
		var me = this;
		if(!me.dialog) {
			me.dialog = new wn.ui.Dialog({
				title: 'Add to To Do',
				width: 350,
				fields: [
					{fieldtype:'Link', fieldname:'assign_to', options:'Profile', 
						label:'Assign To', 
						description:'Add to To Do List of', reqd:true},
					{fieldtype:'Data', fieldname:'description', label:'Comment'}, 
					{fieldtype:'Date', fieldname:'date', label:'Complete By'}, 
					{fieldtype:'Select', fieldname:'priority', label:'Priority',
						options:'Low\nMedium\nHigh', 'default':'Medium'},
					{fieldtype:'Check', fieldname:'notify', label:'Notify By Email'},
					{fieldtype:'Button', label:'Add', fieldname:'add_btn'}
				]
			});
			me.dialog.fields_dict.add_btn.input.onclick = function() {
				
				var assign_to = me.dialog.fields_dict.assign_to.get_value();
				if(assign_to) {
					$c('webnotes.widgets.form.assign_to.add', {
						doctype: me.doctype,
						name: me.name,
						assign_to: assign_to,
						description: me.dialog.fields_dict.description.get_value(),
						priority: me.dialog.fields_dict.priority.get_value(),
						date: me.dialog.fields_dict.date.get_value(),
						notify: me.dialog.fields_dict.notify.get_value()
					}, function(r,rt) {me.render(r.message);});
				}
			}
		}
		me.dialog.clear();
		me.dialog.show();
	}
});

